### ISSUE TITLE
The compiler does not report errors for recursive calls that can not be ended

### ISSUE THREAD

**Issue by hacbit** [rust-lang/rust#128217]
### Code

```Rust
fn foo() {
    bar()
}
fn bar() {
    foo()
}
fn main() {
    foo()
}
```

### Current output

```Shell
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```

### Desired output

```Shell
It shouldn't pass compilation :(
```

### Rationale and extra context

I think the compiler should detect recursive calls that can't end and report an error, rather than exiting at runtime due to a stack overflow. btw, I tried checking with `cargo clippy` but didn't detect the issue

### Rust Version

```Shell
rustc 1.81.0-nightly (3cf924b93 2024-06-15)
binary: rustc
commit-hash: 3cf924b934322fd7b514600a7dc84fc517515346
commit-date: 2024-06-15
host: x86_64-pc-windows-msvc
release: 1.81.0-nightly
LLVM version: 18.1.7
```

**Comment by asquared31415** [comment]
Duplicate of #57965
It is known that indirect recursion (anything not directly calling itself) is not detected at this time.

**Comment by hacbit** [comment]
> Duplicate of #57965 It is known that indirect recursion (anything not directly calling itself) is not detected at this time.

thanks, i get it.
